### ISSUE TITLE
Conditional param's

### ISSUE THREAD

**Issue by david-waterworth** [optuna/optuna#1459]
I just wanted to confirm how one deals with a parameter which has a different distribution depending on the value of another parameter, i.e. in the case below (using CatBoost), I have to limit `max_depth` between 1 and 8 when using Depthwise or Symmetric `grow_policy` but can use 1 to 16 for `Lossguide`

Can I use the same param name for both branches? In the case below I named it differently as I was getting a warning. The problem then though is at the end of training `trial.params` contains the `name` arguments of `suggest_*` not the hparams keys.

        if hparams['grow_policy']=='Lossguide':
            # Max leaves is only used for Lossguide grow_policy
            hparams['max_leaves'] = int(trial.suggest_loguniform("max_leaves", 2, 64))
            hparams['max_depth'] = trial.suggest_int("max_depth_16", 1, 16) # or trial.suggest_int("max_depth", 1, 16) ?
        else:
            hparams['max_depth'] = trial.suggest_int("max_depth_8", 1, 8) # or trial.suggest_int("max_depth", 1, 8) ?

Also, in the case below `grow_policy` must be 'SymmetricTree' when 'boosting_type' == 'Plain', so I didn't sample `grow_policy`  in the `else` branch. Problem then is it's not part of the `trial.params` and I have to apply the same logic to infer the correct hparams for the best trial. Should I use `suggest_categorical` with a single category as per the comment?

        hparams['boosting_type'] = trial.suggest_categorical('boosting_type', ['Plain', 'Ordered'])
        if hparams['boosting_type'] == 'Plain':
            hparams['grow_policy'] = trial.suggest_categorical('grow_policy', ['SymmetricTree', 'Depthwise', 'Lossguide'])
        else:
            # On GPU grow policy Depthwise can't be used with ordered boosting
            # On GPU grow policy Lossguide can't be used with ordered boosting
            hparams['grow_policy'] = 'SymmetricTree' # or trial.suggest_categorical('grow_policy', ['SymmetricTree' ])
            hparams['has_time'] = bool(trial.suggest_int("has_time", 0, 1))

Or am I just doing it wrong, should I be attaching hparams to each trial somehow, since trialparams isn't what I'm actually training the model with? I'm used to hyperopt which had a way of recreating the hparams from the `space` and `params`.

**Comment by HideakiImamura** [comment]
Thanks for the question. 

First, in Optuna learning scheme, we do not recommend giving the same name to the semantically different parameters, because Optuna’s sampler suggests the next parameter based on previous observations filtered by its parameter name. It means that if we give the same name to 2 different parameters, the optimization performance will degrade due to the contamination of the observation history. 

Secondly, we recommend to use a simple substitution (e.g. hparams['grow_policy'] = 'SymmetricTree' in your case) rather than `suggest_categorical` with a single category. It helps us to avoid computational overhead.

**Comment by david-waterworth** [comment]
Thanks for the response

In the second case though  (e.g. hparams['grow_policy'] = 'SymmetricTree') how do you then reconstruct the params after training? i.e. the `study.best_trial.params` will not contain a value for `grow_policy` even if the model was trained with this parameter and it minimises the objective. 

It seems to me the safest thing is to store hparams, i.e. trial.set_user_attr('hparams') (or pickle the actual model) since there's really no reason that trial `study.best_trial.params` and `hparams` have a 1:1 correspondence. This I guess is why hyperopt has a method to map from one to the other?

**Comment by HideakiImamura** [comment]
> In the second case though (e.g. hparams['grow_policy'] = 'SymmetricTree') how do you then reconstruct the hparams after training?

In my understanding, the  meaning of the `hparams['grow_policy']` is different between the different branches,  that is `hparams['boosting_type'] == 'Plain'` or not. So, I think we should not use the same name in Optuna for `hparams['grow_policy']` in the different branches. Then, we can determine the value of `hparams['grow_policy']` based on the `hparams['boosting_type']` after the training.

> It seems to me the safest thing is to store hparams, i.e. trial.set_user_attr('hparams') (or pickle the actual model) since there's really no reason that trial study.best_trial.params and hparams have a 1:1 correspondence. This I guess is why hyperopt has a method to map from one to the other?

Hmm... I guess so too. But, I think if we want to reconstruct hyperparameters after the training, we can just use `study.best_trials.params` and judge the value of `hparams` based on the branch which Optuna selected. Anyway, I recommend not to give the same parameter name to semantically different parameters. (`hparams['grow_policy']`s in different branches looks similar but are actually different.)

**Comment by david-waterworth** [comment]
It's not really that it has a different meaning, there are 3 unique ways catboost can add leaf nodes to a tree but which are implemented depends on other parameters (in this case which type of gradient boosting is used). Many of the options are also conditional on device type as well - i.e. some are implemented only on CPU or GPU. But I agree naming them differently makes sense.

Thinking through it some more, it seems that a good solution is to refactor the code which generates hparams for a trial into a separate function, i.e.

```
def hparams_for_trial(trial):
    hparams={}
    hparams['boosting_type'] = trial.suggest_categorical('boosting_type', ['Plain', 'Ordered'])
    if hparams['boosting_type'] == 'Plain':
        hparams['grow_policy'] = trial.suggest_categorical('grow_policy', ['SymmetricTree', 'Depthwise', 'Lossguide'])
    else:
        # On GPU grow policy Depthwise can't be used with ordered boosting
        # On GPU grow policy Lossguide can't be used with ordered boosting
        hparams['grow_policy'] = 'SymmetricTree' # or trial.suggest_categorical('grow_policy', ['SymmetricTree' ])
        hparams['has_time'] = bool(trial.suggest_int("has_time", 0, 1))
    return hparams
```
I'm making an assumption here - correct me if I'm wrong - but `study.best_trial` is a `FrozenTrial` (I'm typing from memory, that may not be the actual class name). So my second assumption is if you call trial.suggest_* on a frozen trial it should be deterministic - ie. returns the same parameter every time?

Which means after training I can call

```
best_hparams = hparams_for_trial(study.best_trials)
```

What I'm trying to do is avoid creating separate `forward` and `backward` methods and keep them in sync. It's easy to mess up with complex graphs. I like how optuna makes it trivial to write 'forward', it would be even nicer if it could "tape" it like autograd and also generate the 'backward' function :)

**Comment by HideakiImamura** [comment]
> I'm making an assumption here - correct me if I'm wrong - but study.best_trial is a FrozenTrial 

Yes, you are right! It is an instance of `FrozenTrial`.

> So my second assumption is if you call trial.suggest_* on a frozen trial it should be deterministic - ie. returns the same parameter every time?

Unfortunately, the `trial` argument of the `hparams_for_trial` is not an instance of `FrozenTrial` but that of `Trial`. In the current implementation, `FrozenTrial` cannot be used samely as `Trial`, that is, `suggest_*` methods are not implemented for `FrozenTrial`. Therefore, currently, we cannot call `best_hparams = hparams_for_trial(study.best_trial)`. Instead, you can use the [`FixedTrial`](https://optuna.readthedocs.io/en/latest/reference/generated/optuna.trial.FixedTrial.html#optuna.trial.FixedTrial) after training. Please see the reference for detail.
```python
# You have to implement the `to_fixed_from_frozen` function
best_fixed_trial = to_fixed_from_frozen(study.best_trial)

best_hparams = hparams_for_trial*best_fixed_trial)
```

**Comment by HideakiImamura** [comment]
[FYI] We are planning that we make `FrozenTrial` can be used as `FrozenTrial.suggest_*` samely as `Trial.suggest_*`. Ongoing at Issue https://github.com/optuna/optuna/issues/1251.

**Comment by david-waterworth** [comment]
Thanks I was just about to ask that, I'll close now and follow that issue!

**Comment by HideakiImamura** [comment]
Let me close this issue because the feature discussed in #1251 has been merged in #1503.
